src/main/java/com/laytonsmith/core/compiler/signature/SignatureBuilder.java

```diff
@@ -0,0 +1,111 @@
+package com.laytonsmith.core.compiler.signature;
+
+import com.laytonsmith.core.compiler.signature.FunctionSignatures.MatchType;
+import com.laytonsmith.core.constructs.CClassType;
+import com.laytonsmith.core.exceptions.CRE.CREThrowable;
+
+/**
+ * A builder for {@link FunctionSignatures}.
+ * @author P.J.S. Kools
+ */
+public class SignatureBuilder {
+
+	private final FunctionSignatures signatures;
+	private FunctionSignature signature;
+
+	/**
+	 * Creates a new {@link SignatureBuilder}, initialized with a {@link FunctionSignature} with the given return type.
+	 * When determining the return type for given argument types, all matching signatures will be used.
+	 * @param returnType - The return type for the first {@link FunctionSignature}.
+	 */
+	public SignatureBuilder(CClassType returnType) {
+		this(returnType, MatchType.MATCH_ALL);
+	}
+
+	/**
+	 * Creates a new {@link SignatureBuilder}, initialized with a {@link FunctionSignature} with the given return type.
+	 * @param returnType - The return type for the first {@link FunctionSignature}.
+	 * @param matchType - The {@link MatchType} used for determining the return type for given argument types.
+	 */
+	public SignatureBuilder(CClassType returnType, MatchType matchType) {
+		this.signatures = new FunctionSignatures(matchType);
+		this.signature = new FunctionSignature(new ReturnType(returnType));
+	}
+
+	/**
+	 * Adds a normal function parameter. Parameters should be added from left to right.
+	 * @param paramType - The {@link CClassType} of the parameter.
+	 * @param paramName - The name of the parameter.
+	 * @param isOptional - Whether the parameter is optional or not.
+	 * @return This {@link SignatureBuilder}, for chaining builder methods.
+	 */
+	public SignatureBuilder param(CClassType paramType, String paramName, boolean isOptional) {
```

Comment:
Hmm. We will probably want to expand this later to add per-parameter documentation. Especially as the IDE matures, this will become even more useful. I think it's probably worth going ahead and defining this now, since it will be a HUGE hassle to have to go back *again* and backfill this.

Reply:
That's fair. The return type should probably also have documentation then, but `void` would be pointless to document, so there should also be a way around that. I'm wondering whether we should make per-param and return type docs mandatory, given that they can also be explained in the general function docs, which in Java often results in duplicate information. I think that it would be wise to at least add an option to add parameter and return type documentation, but not require it (yet).

Reply:
It's fine to accept java null, to mean "no documentation here" but it needs to be explicit.

Reply:
Added per-param docs.